File: python/src/agent_api/local/workdir.py
from __future__ import annotations

import hashlib
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .errors import LocalError, LocalFileTooLargeError, LocalIgnoredPathError, LocalNotTextFileError
from .files import LocalFileStore
from .paths import (
    classify_local_path_sensitivity,
    default_workdir_ignore_rules,
    entry_from_stat,
    ignored,
    line_edit_params,
    looks_binary,
    normalize_relative_path,
    parse_ignore_file,
    patch_line_range,
    positive_int,
    snapshot_file_changed,
    split_lines,
)
from .types import LocalFileStat, LocalIgnoreRule


class LocalWorkdirManager:
    def open(self, root: str | Path, **options: Any) -> "LocalWorkdir":
        return LocalWorkdir(root, **options)


@dataclass
class LocalWorkdir:
    root: Path | str
    name: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)
    trusted: bool = False
    ignore: list[LocalIgnoreRule] = field(default_factory=list)
    gitignore: bool = True
    max_file_bytes: int = 10 * 1024 * 1024

    def __post_init__(self) -> None:
        self.root = Path(self.root).expanduser().resolve()
        self.name = (self.name or self.root.name or "workdir").strip()
        self.files = LocalFileStore(self.root, label="workdir")
        self.ignore = [*default_workdir_ignore_rules(), *self.ignore]

    def ensure(self) -> None:
        self.files.ensure()
        if self.gitignore:
            self.load_ignore_files()

    def load_ignore_files(self, files: list[str] | None = None) -> list[LocalIgnoreRule]:
        loaded: list[LocalIgnoreRule] = []
        for file in files or [".gitignore"]:
            try:
                loaded.extend(parse_ignore_file(self.files.read_text(file)))
            except FileNotFoundError:
                continue
        self.ignore.extend(loaded)
        return loaded

    def child(self, relative_path: str | Path, **options: Any) -> "LocalWorkdir":
        return LocalWorkdir(
            self.files.resolve_path(relative_path),
            name=options.get("name"),
            metadata=options.get("metadata", self.metadata),
            trusted=options.get("trusted", self.trusted),
            ignore=options.get("ignore", self.ignore),
            gitignore=options.get("gitignore", self.gitignore),
            max_file_bytes=options.get("max_file_bytes", self.max_file_bytes),
        )

    def resolve_path(self, relative_path: str | Path = ".") -> Path:
        self._assert_allowed(relative_path)
        return self.files.resolve_path(relative_path)

    def list(self, relative_path: str | Path = ".", **options: Any) -> list[LocalFileStat]:
        options["ignore"] = self._merge_ignore(options.get("ignore"))
        return self.files.list(relative_path, **options)

    def list_entries(self, relative_path: str | Path = ".", **options: Any) -> dict[str, Any]:
        options["ignore"] = self._merge_ignore(options.get("ignore"))
        return self.files.list_entries(relative_path, **options)

    def search_entries(self, **params: Any) -> dict[str, Any]:
        query = str(params.get("query", "")).strip().lower()
        if not query:
            raise ValueError("query is required")
        limit = positive_int(params.get("limit"), 100)
        path = params.get("path", ".")
        return {
            "object": "list",
            "entries": [
                entry_from_stat(item)
                for item in self.list(path, recursive=True, include_directories=True)
                if query in item.path.lower()
            ][:limit],
        }

    def read_file(self, relative_path: str | Path, **params: Any) -> dict[str, Any]:
        self._assert_allowed(relative_path)
        params.setdefault("max_bytes", self.max_file_bytes)
        return self.files.read_file(relative_path, **params)

    def write_file(self, relative_path: str | Path, content: str | bytes, **options: Any) -> dict[str, Any]:
        self._assert_allowed(relative_path)
        return self.files.write_file(relative_path, content, **options)

    def read_text(self, relative_path: str | Path) -> str:
        self._assert_allowed(relative_path)
        return self.files.read_text(relative_path)

    def write_text(self, relative_path: str | Path, content: str, **options: Any) -> Path:
        self._assert_allowed(relative_path)
        return self.files.write_text(relative_path, content, **options)

    def delete_path(self, relative_path: str | Path) -> dict[str, Any]:
        self._assert_allowed(relative_path)
        return self.files.delete_path(relative_path)

    def create_directory(self, relative_path: str | Path = ".") -> dict[str, Any]:
        self._assert_allowed(relative_path)
        return self.files.create_directory(relative_path)

    def read_lines(self, relative_path: str | Path, **params: Any) -> dict[str, Any]:
        self._assert_allowed(relative_path)
        params.setdefault("max_bytes", self.max_file_bytes)
        return self.files.read_lines(relative_path, **params)

    def preview_patch_lines(self, relative_path: str | Path, **params: Any) -> dict[str, Any]:
        self._assert_allowed(relative_path)
        lines = self.read_lines(
            relative_path,
            start_line=params["start_line"],
            end_line=params.get("end_line"),
            max_bytes=params.get("max_bytes", self.max_file_bytes),
        )
        file = self.files.read_file(relative_path, max_bytes=params.get("max_bytes", self.max_file_bytes), format="raw")
        if file["truncated"]:
            raise LocalFileTooLargeError(str(relative_path))
        if looks_binary(file["content"]):
            raise LocalNotTextFileError(str(relative_path))
        _, _, total = patch_line_range(file["content"].decode("utf-8"), params["start_line"], params.get("end_line"), params.get("replacement", ""))
        return {
            "path": lines["path"],
            "start_line": lines["start_line"],
            "end_line": lines["end_line"],
            "total_lines": total,
            "before": lines["lines"],
            "after": [] if params.get("replacement", "") == "" else split_lines(params.get("replacement", "")),
        }

    def patch_lines(self, relative_path: str | Path, **params: Any) -> dict[str, Any]:
        self._assert_allowed(relative_path)
        params.setdefault("max_bytes", self.max_file_bytes)
        return self.files.patch_lines(relative_path, **params)

    def preview_edits(self, edits: list[dict[str, Any]]) -> dict[str, Any]:
        previews = []
        for edit in edits:
            self._assert_expected_hash(edit["path"], edit.get("expected_sha256"))
            previews.append(self.preview_patch_lines(edit["path"], **line_edit_params(edit)))
        return {"edits": [dict(edit) for edit in edits], "previews": previews}

    def apply_edits(self, edits: list[dict[str, Any]]) -> dict[str, Any]:
        backups: list[dict[str, str]] = []
        applied: list[dict[str, Any]] = []
        try:
            for edit in edits:
                self._assert_expected_hash(edit["path"], edit.get("expected_sha256"))
                backups.append({"path": edit["path"], "content": self.read_text(edit["path"])})
                applied.append(self.patch_lines(edit["path"], **line_edit_params(edit)))
        except Exception:
            for backup in reversed(backups):
                self.write_text(backup["path"], backup["content"])
            raise
        return {"applied": applied, "backups": backups}

    def classify_path(self, relative_path: str | Path) -> dict[str, Any]:
        return classify_local_path_sensitivity(relative_path)

    def grep(self, **params: Any) -> dict[str, Any]:
        params["ignore"] = self._merge_ignore(params.get("ignore"))
        return self.files.grep(**params)

    def summarize(self, **params: Any) -> dict[str, Any]:
        params["ignore"] = self._merge_ignore(params.get("ignore"))
        return self.files.summarize(**params)

    def snapshot(self, *, path: str | Path = ".", hash: bool = True, max_bytes_per_file: int | None = None) -> dict[str, Any]:
        max_bytes = positive_int(max_bytes_per_file, self.max_file_bytes)
        files = []
        for item in self.list(path, recursive=True):
            if item.type != "file":
                continue
            snap: dict[str, Any] = {"path": item.path, "size": item.size, "modified_at_unix": int(item.modified_at)}
            if hash and item.size <= max_bytes:
                snap["sha256"] = hashlib.sha256(Path(item.full_path).read_bytes()).hexdigest()
            files.append(snap)
        return {"root": str(self.root), "name": self.name, "generated_at_unix": int(time.time()), "files": files}

    def diff(self, before: dict[str, Any], after: dict[str, Any]) -> dict[str, Any]:
        before_by_path = {file["path"]: file for file in before.get("files", [])}
        after_by_path = {file["path"]: file for file in after.get("files", [])}
        added = []
        modified = []
        unchanged = []
        for after_file in after_by_path.values():
            before_file = before_by_path.get(after_file["path"])
            if before_file is None:
                added.append(after_file)
            elif snapshot_file_changed(before_file, after_file):
                modified.append({"before": before_file, "after": after_file})
            else:
                unchanged.append(after_file)
        deleted = [file for path, file in before_by_path.items() if path not in after_by_path]
        return {"added": added, "modified": modified, "deleted": deleted, "unchanged": unchanged}

    def _merge_ignore(self, ignore: list[LocalIgnoreRule] | None) -> list[LocalIgnoreRule]:
        return [*self.ignore, *(ignore or [])]

    def _assert_allowed(self, relative_path: str | Path) -> None:
        rel = normalize_relative_path(relative_path)
        if rel != "." and ignored(rel, self.ignore):
            raise LocalIgnoredPathError(rel)

    def _assert_expected_hash(self, relative_path: str | Path, expected_sha256: str | None) -> None:
        if not expected_sha256:
            return
        actual = hashlib.sha256(self.files.resolve_path(relative_path).read_bytes()).hexdigest()
        if actual != expected_sha256:
            raise LocalError("local_edit_conflict", f"local file changed before edit: {relative_path}", path=str(relative_path))

File: go/agentapi/local/workdir.go
package local

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"path/filepath"
	"time"
)

type WorkdirManager struct{}

type WorkdirOptions struct {
	Name             string
	Metadata         map[string]any
	Trusted          bool
	Ignore           []IgnoreRule
	DisableGitignore bool
	MaxFileBytes     int
}

type Workdir struct {
	Root         string
	Name         string
	Metadata     map[string]any
	Trusted      bool
	Files        *FileStore
	Ignore       []IgnoreRule
	Gitignore    bool
	MaxFileBytes int
}

type SnapshotParams struct {
	Path            string
	OmitHash        bool
	MaxBytesPerFile int
}

type SnapshotFile struct {
	Path           string `json:"path"`
	Size           int64  `json:"size"`
	ModifiedAtUnix int64  `json:"modified_at_unix,omitempty"`
	SHA256         string `json:"sha256,omitempty"`
}

type Snapshot struct {
	Root            string         `json:"root"`
	Name            string         `json:"name"`
	GeneratedAtUnix int64          `json:"generated_at_unix"`
	Files           []SnapshotFile `json:"files"`
}

type Diff struct {
	Added     []SnapshotFile `json:"added"`
	Modified  []DiffModified `json:"modified"`
	Deleted   []SnapshotFile `json:"deleted"`
	Unchanged []SnapshotFile `json:"unchanged"`
}

type DiffModified struct {
	Before SnapshotFile `json:"before"`
	After  SnapshotFile `json:"after"`
}

type LinePatchPreview struct {
	Path       string   `json:"path"`
	StartLine  int      `json:"start_line"`
	EndLine    int      `json:"end_line"`
	TotalLines int      `json:"total_lines"`
	Before     []string `json:"before"`
	After      []string `json:"after"`
}

type LineEdit struct {
	Path           string
	StartLine      int
	EndLine        int
	Replacement    string
	ExpectedSHA256 string
}

type EditPlan struct {
	Edits    []LineEdit         `json:"edits"`
	Previews []LinePatchPreview `json:"previews"`
}

type EditResult struct {
	Applied      []FileLinesPatch `json:"applied"`
	ChangedFiles []string         `json:"changed_files"`
	EditCount    int              `json:"edit_count"`
}

type EditBackup struct {
	Path    string `json:"path"`
	Content string `json:"content"`
}

func (m *WorkdirManager) Open(root string, opts WorkdirOptions) (*Workdir, error) {
	return NewWorkdir(root, opts)
}

func NewWorkdir(root string, opts WorkdirOptions) (*Workdir, error) {
	abs, err := filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	name := opts.Name
	if name == "" {
		name = filepath.Base(abs)
	}
	if name == "" || name == "." {
		name = "workdir"
	}
	files, err := NewFileStore(abs, "workdir")
	if err != nil {
		return nil, err
	}
	ignore := append([]IgnoreRule{}, DefaultWorkdirIgnoreRules()...)
	ignore = append(ignore, opts.Ignore...)
	maxBytes := opts.MaxFileBytes
	if maxBytes <= 0 {
		maxBytes = 10 * 1024 * 1024
	}
	return &Workdir{Root: abs, Name: name, Metadata: opts.Metadata, Trusted: opts.Trusted, Files: files, Ignore: ignore, Gitignore: !opts.DisableGitignore, MaxFileBytes: maxBytes}, nil
}

func (w *Workdir) Ensure() error {
	if err := w.Files.Ensure(); err != nil {
		return err
	}
	if w.Gitignore {
		_, err := w.LoadIgnoreFiles(".gitignore")
		return err
	}
	return nil
}

func (w *Workdir) LoadIgnoreFiles(files ...string) ([]IgnoreRule, error) {
	if len(files) == 0 {
		files = []string{".gitignore"}
	}
	var loaded []IgnoreRule
	for _, file := range files {
		text, err := w.Files.ReadText(file)
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return nil, err
		}
		loaded = append(loaded, ParseIgnoreFile(text)...)
	}
	w.Ignore = append(w.Ignore, loaded...)
	return loaded, nil
}

func (w *Workdir) ResolvePath(rel string) (string, error) {
	if err := w.assertAllowed(rel); err != nil {
		return "", err
	}
	return w.Files.ResolvePath(rel)
}

func (w *Workdir) List(rel string, opts ListOptions) ([]FileStat, error) {
	opts.Ignore = w.mergeIgnore(opts.Ignore)
	return w.Files.List(rel, opts)
}

func (w *Workdir) ListWithWarnings(rel string, opts ListOptions) ([]FileStat, []ScanWarning, error) {
	opts.Ignore = w.mergeIgnore(opts.Ignore)
	return w.Files.ListWithWarnings(rel, opts)
}

func (w *Workdir) ListEntries(rel string, opts ListOptions) (*EntryList, error) {
	opts.Ignore = w.mergeIgnore(opts.Ignore)
	return w.Files.ListEntries(rel, opts)
}

func (w *Workdir) SearchEntries(query, rel string, limit int) (*EntryList, error) {
	stats, err := w.List(rel, ListOptions{Recursive: true, IncludeDirectories: true})
	if err != nil {
		return nil, err
	}
	if limit <= 0 {
		limit = 100
	}
	var entries []Entry
	for _, item := range stats {
		if containsFold(item.Path, query) {
			entries = append(entries, entryFromStat(item))
			if len(entries) >= limit {
				break
			}
		}
	}
	return &EntryList{Object: "list", Entries: entries}, nil
}

func (w *Workdir) ReadFile(rel string, params ReadFileParams) (*FileDeliver, error) {
	if err := w.assertAllowed(rel); err != nil {
		return nil, err
	}
	if params.MaxBytes <= 0 {
		params.MaxBytes = w.MaxFileBytes
	}
	return w.Files.ReadFile(rel, params)
}

func (w *Workdir) ReadFileRaw(rel string, params ReadFileParams) (*FileRaw, error) {
	if err := w.assertAllowed(rel); err != nil {
		return nil, err
	}
	if params.MaxBytes <= 0 {
		params.MaxBytes = w.MaxFileBytes
	}
	return w.Files.ReadFileRaw(rel, params)
}

func (w *Workdir) WriteText(rel, content string) (string, error) {
	if err := w.assertAllowed(rel); err != nil {
		return "", err
	}
	return w.Files.WriteText(rel, content)
}

func (w *Workdir) ReadText(rel string) (string, error) {
	if err := w.assertAllowed(rel); err != nil {
		return "", err
	}
	return w.Files.ReadText(rel)
}

func (w *Workdir) WriteFile(rel string, content []byte) (*FileWrite, error) {
	if err := w.assertAllowed(rel); err != nil {
		return nil, err
	}
	return w.Files.WriteFile(rel, content)
}

func (w *Workdir) DeletePath(rel string) (*PathDelete, error) {
	if err := w.assertAllowed(rel); err != nil {
		return nil, err
	}
	return w.Files.DeletePath(rel)
}

func (w *Workdir) CreateDirectory(rel string) (map[string]string, error) {
	if err := w.assertAllowed(rel); err != nil {
		return nil, err
	}
	return w.Files.CreateDirectory(rel)
}

func (w *Workdir) ReadLines(rel string, params ReadLinesParams) (*FileLines, error) {
	if err := w.assertAllowed(rel); err != nil {
		return nil, err
	}
	if params.MaxBytes <= 0 {
		params.MaxBytes = w.MaxFileBytes
	}
	return w.Files.ReadLines(rel, params)
}

func (w *Workdir) PreviewPatchLines(rel string, params PatchLinesParams) (*LinePatchPreview, error) {
	lines, err := w.ReadLines(rel, ReadLinesParams{StartLine: params.StartLine, EndLine: params.EndLine, MaxBytes: firstPositiveInt(params.MaxBytes, w.MaxFileBytes)})
	if err != nil {
		return nil, err
	}
	file, err := w.ReadFileRaw(rel, ReadFileParams{MaxBytes: firstPositiveInt(params.MaxBytes, w.MaxFileBytes)})
	if err != nil {
		return nil, err
	}
	if file.Truncated {
		return nil, fileTooLargeError(rel)
	}
	patched, _, total, err := patchLineRange(string(file.Content), params.StartLine, params.EndLine, params.Replacement)
	if err != nil {
		return nil, err
	}
	_ = patched
	after := []string{}
	if params.Replacement != "" {
		after = splitLines(params.Replacement)
	}
	return &LinePatchPreview{Path: lines.Path, StartLine: lines.StartLine, EndLine: lines.EndLine, TotalLines: total, Before: lines.Lines, After: after}, nil
}

func (w *Workdir) PatchLines(rel string, params PatchLinesParams) (*FileLinesPatch, error) {
	if err := w.assertAllowed(rel); err != nil {
		return nil, err
	}
	if params.MaxBytes <= 0 {
		params.MaxBytes = w.MaxFileBytes
	}
	return w.Files.PatchLines(rel, params)
}

func (w *Workdir) PreviewEdits(edits []LineEdit) (*EditPlan, error) {
	previews := make([]LinePatchPreview, 0, len(edits))
	for _, edit := range edits {
		if err := w.assertExpectedHash(edit.Path, edit.ExpectedSHA256); err != nil {
			return nil, err
		}
		preview, err := w.PreviewPatchLines(edit.Path, PatchLinesParams{StartLine: edit.StartLine, EndLine: edit.EndLine, Replacement: edit.Replacement})
		if err != nil {
			return nil, err
		}
		previews = append(previews, *preview)
	}
	return &EditPlan{Edits: append([]LineEdit(nil), edits...), Previews: previews}, nil
}

func (w *Workdir) ApplyEdits(edits []LineEdit) (*EditResult, error) {
	var backups []EditBackup
	var applied []FileLinesPatch
	for _, edit := range edits {
		if err := w.assertExpectedHash(edit.Path, edit.ExpectedSHA256); err != nil {
			restoreBackups(w, backups)
			return nil, err
		}
		content, err := w.ReadText(edit.Path)
		if err != nil {
			restoreBackups(w, backups)
			return nil, err
		}
		backups = append(backups, EditBackup{Path: edit.Path, Content: content})
		patch, err := w.PatchLines(edit.Path, PatchLinesParams{StartLine: edit.StartLine, EndLine: edit.EndLine, Replacement: edit.Replacement})
		if err != nil {
			restoreBackups(w, backups)
			return nil, err
		}
		applied = append(applied, *patch)
	}
	return &EditResult{Applied: applied, ChangedFiles: uniquePatchPaths(applied), EditCount: len(applied)}, nil
}

func (w *Workdir) ClassifyPath(rel string) PathSensitivityInfo {
	return ClassifyPathSensitivity(rel)
}

func (w *Workdir) Grep(params GrepParams) (*GrepResponse, error) {
	params.Ignore = w.mergeIgnore(params.Ignore)
	return w.Files.Grep(params)
}

func (w *Workdir) Summarize(params SummaryParams) (*Summary, error) {
	params.Ignore = w.mergeIgnore(params.Ignore)
	return w.Files.Summarize(params)
}

func (w *Workdir) Snapshot(params SnapshotParams) (*Snapshot, error) {
	path := firstNonEmpty(params.Path, ".")
	maxBytes := firstPositiveInt(params.MaxBytesPerFile, w.MaxFileBytes)
	hashFiles := !params.OmitHash
	stats, err := w.List(path, ListOptions{Recursive: true})
	if err != nil {
		return nil, err
	}
	out := &Snapshot{Root: w.Root, Name: w.Name, GeneratedAtUnix: time.Now().Unix()}
	for _, item := range stats {
		if item.Type != FileTypeFile {
			continue
		}
		file := SnapshotFile{Path: item.Path, Size: item.Size, ModifiedAtUnix: item.ModifiedAt.Unix()}
		if hashFiles && item.Size <= int64(maxBytes) {
			raw, err := os.ReadFile(item.FullPath)
			if err != nil {
				return nil, err
			}
			sum := sha256.Sum256(raw)
			file.SHA256 = hex.EncodeToString(sum[:])
		}
		out.Files = append(out.Files, file)
	}
	return out, nil
}

func (w *Workdir) Diff(before, after *Snapshot) Diff {
	beforeByPath := map[string]SnapshotFile{}
	afterByPath := map[string]SnapshotFile{}
	for _, file := range before.Files {
		beforeByPath[file.Path] = file
	}
	for _, file := range after.Files {
		afterByPath[file.Path] = file
	}
	var diff Diff
	for _, file := range after.Files {
		old, ok := beforeByPath[file.Path]
		if !ok {
			diff.Added = append(diff.Added, file)
		} else if snapshotFileChanged(old, file) {
			diff.Modified = append(diff.Modified, DiffModified{Before: old, After: file})
		} else {
			diff.Unchanged = append(diff.Unchanged, file)
		}
	}
	for _, file := range before.Files {
		if _, ok := afterByPath[file.Path]; !ok {
			diff.Deleted = append(diff.Deleted, file)
		}
	}
	return diff
}

func (w *Workdir) mergeIgnore(extra []IgnoreRule) []IgnoreRule {
	out := append([]IgnoreRule{}, w.Ignore...)
	out = append(out, extra...)
	return out
}

func (w *Workdir) assertAllowed(rel string) error {
	clean, err := NormalizeRelativePath(rel)
	if err != nil {
		return err
	}
	if clean != "." && Ignored(clean, w.Ignore) {
		return ignoredPathError(clean)
	}
	return nil
}

func (w *Workdir) assertExpectedHash(rel, expected string) error {
	if expected == "" {
		return nil
	}
	full, err := w.Files.ResolvePath(rel)
	if err != nil {
		return err
	}
	raw, err := os.ReadFile(full)
	if err != nil {
		return err
	}
	sum := sha256.Sum256(raw)
	if hex.EncodeToString(sum[:]) != expected {
		return editConflictError(rel)
	}
	return nil
}

func restoreBackups(w *Workdir, backups []EditBackup) {
	for i := len(backups) - 1; i >= 0; i-- {
		_, _ = w.WriteText(backups[i].Path, backups[i].Content)
	}
}

func uniquePatchPaths(patches []FileLinesPatch) []string {
	seen := map[string]bool{}
	var paths []string
	for _, patch := range patches {
		if seen[patch.Path] {
			continue
		}
		seen[patch.Path] = true
		paths = append(paths, patch.Path)
	}
	return paths
}
